design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,306 @@
+
+### Goals
+- Define the process a feature needs to pass for graduation.
+- Define the process a feature needs to pass in order to be removed.
+- Provide policies and rules on how to manage a feature during its
+  lifetime.
+
+### Non Goals
+- Implement enforcement tooling to keep features in sync with
+  the lifecycle rules.
+
+### Definition Of Users
+- Development contributors.
+- Cluster operators.
+
+### User Stories
+- As a KubeVirt contributor, I would like to introduce a new useful
+  feature and follow it to graduation (GA).
+- As a KubeVirt contributor, I would like to remove a feature
+  that has not yet reached its formal graduation.
+- As a KubeVirt contributor, I would like to remove a feature
+  that has already graduated.
+  > **Note**: Removal of a GA feature is considered an exception.
+  > Strong arguments and a wide agreement is required for such an action.
+- As a KubeVirt cluster operator, I would like to experiment with
+  a newly-proposed ("Alpha") feature in a controlled environment, to see
+  if it makes sense to me.
+- As a KubeVirt cluster operator, I would like to evaluate an
+  undergraduate ("Beta") feature in a real-life environment with actual users.
+- As a KubeVirt cluster operator, I would like to keep using a feature
+  that got graduated after I used it during the evaluation period.
+- As a KubeVirt cluster operator, I would like to know that
+  an undergraduate feature is planned to be removed.
+- As a KubeVirt cluster operator, I would like to stop using a feature
+  that got removed.
+
+### Repos
+This is a cross repo project policy under the
+[kubevirt](https://github.com/kubevirt) organization.
+
+## Proposal Design
+The proposal on how to define a feature lifecycle is influenced by
+processes and policies from the Kubernetes project.
+These sources are scattered around, each focusing on different
+aspects of a feature:
+- [Feature Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)
+- [Graduation](https://kubernetes.io/blog/2020/08/21/moving-forward-from-beta/)
+- [Changing the API](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md)
+- [Deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)
+
+The proposal takes the top-down approach, starting with the high level
+flow that a common feature will traverse through.
+Continuing with actions that need to be taken and timeline suggestions.
+
+Both feature graduation and discontinuation flows are covered.
+Including the implications on users.
+
+Depending on individual topics, follow-up proposals may extend the basic
+points raised in this proposal.
+
+### Feature Stages
+A feature is expected to pass through the following stages:
+1. Enhancement proposal.
+2. Implementation.
+3. Release as Alpha (experimental).
+4. Release as Beta (pre-release for evaluation).
+5. Release as General Availability (graduation).
+6. Removal.
```

Comment:
Should this list include optional and mandatory steps?
Specifically removal is optional.

But maybe this is obvious

Reply:
Alpha could also be optional at times.

Reply:
This itself is worth noting. Maybe something like "(while the list is ordered, the steps can be skipped, e.g skipping Alpha)".